### DISCUSSION TITLE
Using Redis with tagging

### DISCUSSION THREAD

**Discussion by mattzarb** [symfony/symfony#43508] [Q&A] [answered]
trying to use Redis with tagging in my Symfony 5 app but can't seem to get RedisTagAwareAdapter to work. Saving to Redis without tags works just fine like this:

```
use Symfony\Component\Cache\Adapter\RedisAdapter;
use Symfony\Contracts\Cache\ItemInterface;

$client = RedisAdapter::createConnection(
    'redis://redis'
);
$cache = new RedisAdapter($client);

$key = "testkey";
$data = "hello world";

$cacheItem = $cache->getItem($key);
$cacheItem->set($data);
$cacheItem->expiresAfter(3600);
$cache->save($cacheItem);
```

But if I switch to using the RedisTagAwareAdapter as [this suggests](https://symfony.com/doc/current/components/cache/adapters/redis_adapter.html#working-with-tags) then nothing gets saved:

```
use Symfony\Component\Cache\Adapter\RedisAdapter;
use Symfony\Component\Cache\Adapter\RedisTagAwareAdapter;
use Symfony\Contracts\Cache\ItemInterface;

$client = RedisAdapter::createConnection(
    'redis://redis'
);
$cache = new RedisTagAwareAdapter($client);

$key = "testkey";
$data = "hello world";

$cacheItem = $cache->getItem($key);
$cacheItem->set($data);
$cacheItem->tag('greeting');
$cacheItem->expiresAfter(3600);
$cache->save($cacheItem);
```

the $cache->save() returns false. No other errors are thrown.

I'm using Symfony 5.1.11, Redis server version 6.2.5 and have phpredis installed. Any ideas on what I'm doing wrong? TIA

**Comment by michaoj** [comment, marked as answer]
This could not work as the cache aware adapter has been introduced in sf 5.2 only.

![image](https://user-images.githubusercontent.com/4163887/137332827-d17ae6d7-a4f0-4d84-b39e-b3a64328f109.png)

Now, considering you are able to update to 5.2, maybe I'm off topic, but why do you need to declare your cache connection explicitly?

If you use the cache section of the framework.yaml config file, you can achieve quite the same thing by using a dedicated cache pool
```yaml
framework:
   cache:
        pools:
            mytag.cache:
                adapter: cache.adapter.redis_tag_aware
                provider: '%env(REDIS_URL)%'
                default_lifetime: '%env(REDIS_CACHE_KEY_LIFETIME)%'
                tags: true
```
You can inject the cache by using
```php
public function __construct(
        private CacheInterface $mytagCache,
    ) {
}
```
And would be able to get/set a tagged item with
```php
$this->mytafCache->get(
            'mykey',
            function (ItemInterface $item) {
                $item->tag('The tag I want');
                ...generate data...
            }
        );
```
Please not that this code is for PHP8, so you may have to adapt the construct if you're on an older version of PHP

**Reply by mattzarb** [reply]
Excellent, thanks for the information. I completely glossed over the 5.2 number. Wrt the dedicated cache pools, that was on my to-do list, I was just trying to simplify my code for this post. Thanks much.

**Reply by mattzarb** [reply]
@michaoj Perhaps you can answer this one for me as well. I'm trying to set up my cache pool as you described above, so that I can use this in an event subscriber:

**framework.yaml:**
```
framework:
    cache:
        pools:
            itc.cache:
                adapter: cache.adapter.redis
                provider: redis://redis
                default_lifetime: 3600
```

**CacheEventSubscriber.php:**
```
namespace App\EventSubscriber;

use Symfony\Component\HttpKernel\KernelEvents;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Security\Core\Security;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Contracts\Cache\CacheInterface;

class CacheEventSubscriber implements EventSubscriberInterface
{
    private Security $security;

    private CacheInterface $cache;

    /**
     * @param Security $security
     */
    public function __construct(Security $security, CacheInterface $itcCache)
    {
        $this->security = $security;
        $this->cache = $itcCache;
    }

    /**
     * @return array
     */
    public static function getSubscribedEvents()
    {
        return [
            KernelEvents::RESPONSE => 'cacheResponse',
        ];
    }

    public function cacheResponse(ResponseEvent $event)
    {
        if ($event->isMasterRequest()) {
            $response = $event->getResponse();
            $request = $event->getRequest();
            if ($request->getMethod() == 'GET') {

                $key = preg_replace('/[^a-z0-9]+/', '_', strtolower($request->getRequestUri()));

                $cacheItem = $this->cache->getItem($key);

                if (!$cacheItem->isHit()) {

                    $statusCode = $response->getStatusCode();
                    if ($statusCode == 200) {
                        $cacheItem->set(serialize($response->getContent()));
                        $cacheItem->expiresAfter(3600);
                        $this->cache->save($cacheItem);
                    }

                }
            }
        }
    }
}
```
The request works and I get no errors but nothing is showing in the cache. Am I doing something wrong in my implementation?

**Reply by michaoj** [reply]
Are you sure you reach your subscriber ? Maybe you can use a dump, dd or a debugger to check this?

and why do you use isHit to generate your key ? The getCache method takes an anonymous function as second parameter which allo you to define the key content if it's not found.

**Reply by mattzarb** [reply]
I've already used dump to make sure this is being reached. In fact, if I put in my explicit connection from my original post everything works as expected. I can see content saved in Redis. It's only when I pulled that out and tried to use the cache pools with dependency injection that it failed.

With respect to the anonymous function, I saw it being done both ways in the documentation which I found confusing. If you [look here](https://symfony.com/doc/current/components/cache/cache_items.html#cache-item-hits-and-misses) you'll see it the way I'm doing it. Not sure what the difference is - something to do with PSR6 maybe? But as I said, this was working until I removed the explicit connection. I'll try it as an anonymous function and see what happens with that.

**Reply by michaoj** [reply]
Try to quote your provider and check you do not have any errors in your log
